BinaryEncoding.md

```diff
@@ -47,22 +49,69 @@ where the former two are used for compatibility with potential future extensions
 A [Signed LEB128](https://en.wikipedia.org/wiki/LEB128#Signed_LEB128) variable-length integer, limited to _N_ bits (i.e., the values [-2^(_N_-1), +2^(_N_-1)-1]),
 represented by _at most_ ceil(_N_/7) bytes that may contain padding `0x80` or `0xFF` bytes.
 
-Note: Currently, the only sizes used are `varint32` and `varint64`.
+Note: Currently, the only sizes used are `varint7`, `varint32` and `varint64`.
+
+## Language Types
+
+Note: All types are represented by negative `varint7` values. This is so that they can coexist in a single space with (positive) indices into the type section, which may be relevant in future extensions of the type system.
 
 ### `value_type`
-A single-byte unsigned integer indicating a [value type](Semantics.md#types). These types are encoded as:
-* `1` indicating type `i32` 
-* `2` indicating type `i64` 
-* `3` indicating type `f32` 
-* `4` indicating type `f64`
-
-### `inline_signature_type`
-A single-byte unsigned integer indicating a signature. These types are encoded as:
-* `0` indicating a signature with 0 results.
-* `1` indicating a signature with 1 result of type `i32`.
-* `2` indicating a signature with 1 result of type `i64`.
-* `3` indicating a signature with 1 result of type `f32`.
-* `4` indicating a signature with 1 result of type `f64`.
+A `varint7` indicating a [value type](Semantics.md#types). These types are encoded as:
+* `-0x01` (i.e., the byte `0x7f`) indicating type `i32`
+* `-0x02` (i.e., the byte `0x7e`) indicating type `i64`
+* `-0x03` (i.e., the byte `0x7d`) indicating type `f32`
+* `-0x04` (i.e., the byte `0x7c`) indicating type `f64`
+
+### `block_type`
+A `varint7` indicating a signature. These types are encoded as:
+* `-0x40` (i.e., the byte `0x40`) indicating a signature with 0 results.
```

Comment:
Hm, by starting with the lowest absolute value we keep an open-ended gap at the "end" for future extension. The inverse order would become rather weird if we were to grow out of the 6 bit negative range of a varint7, because then we'd have to wrap around and add additional types or categories "before" the basic types.